### PR TITLE
Cai2Hcl: refactor of converter_map and split of utils.go

### PR DESCRIPTION
**Pull request by amirkaromashkin**
Refactor converter_map.go to converter.go, which implements Converter interface.

Introduce UberConverter implementation which allows to implement a Converter for the whole service (i.e. compute) folder.

This PR does not change any logic.

```release-note:none
```

### CODE REVIEW

File: .ci/scripts/build-environment/downstream-builder/generate_downstream.sh [hunk 1]
```diff
@@ -120,16 +120,12 @@ if [ "$REPO" == "terraform-google-conversion" ]; then
 
     bundle exec compiler.rb -a -e terraform -f validator -o $LOCAL_PATH/tfplan2cai -v $VERSION
 
-    # Ignore 'tgc-base' step, because this is a new provider.
-    # TODO: remove this condition on next cai2hcl-related PR.
```
[the remark this answers is not in the document]
Comment (melinath) [reply .ci/scripts/build-environment/downstream-builder/generate_downstream.sh:124]: It looks to me like this could be removed in a separate PR. I think it would be better to do it in a standalone PR _before_ this PR so that this PR will show an accurate diff.

[the remark this answers is not in the document]
Comment (amirkaromashkin) [reply .ci/scripts/build-environment/downstream-builder/generate_downstream.sh:124]: Sure, here is a separate PR for this line change: https://github.com/GoogleCloudPlatform/magic-modules/pull/9407.

Will update this PR once merged.

[the remark this answers is not in the document]
Comment (amirkaromashkin) [reply .ci/scripts/build-environment/downstream-builder/generate_downstream.sh]: Done

**Comment by rileykarson** [comment]
Thanks! Working through this.

**Review by melinath — CHANGES_REQUESTED** [review]
This is a pretty massive set of changes - could you split it up into multiple PRs? I'd suggest at a minimum: one PR for the converter / shared files refactor and one for each resource. That should make the overall review go faster. It would probably also be useful to provide some context about why the changes to the converter / shared files are required.

**Comment by amirkaromashkin** [comment]
> This is a pretty massive set of changes - could you split it up into multiple PRs? I'd suggest at a minimum: one PR for the converter / shared files refactor and one for each resource. That should make the overall review go faster. It would probably also be useful to provide some context about why the changes to the converter / shared files are required.

Right, it was not that big of a PR when Ruby generator was there, but in any case you're right, too many logical changes. It can be split into 3 chunks:
1. Refactor of converter_map and split of utils.go
2. Manually add ForwardingRule generated converter 
3. Ability to match converters by ResourceName + Manually add BackendService+GlobalBackendService generated converter 

So I have updated this PR to Part1: Refactor of converter_map and split of utils.go.
PTAL. Thanks!

**Comment by rileykarson** [comment]
@melinath do you mind picking up review? I've been unable to fit this in since I'll need to onboard myself to TGC enough to review in the process.

**Comment by melinath** [comment]
yep, can do!

File: mmv1/third_party/cai2hcl/services/compute/compute_forwarding_rule_test.go [hunk 2]
```diff
@@ -1,17 +1,29 @@
...
+// ----------------------------------------------------------------------------
+//
+//     This file is automatically generated by Magic Modules and manual
+//     changes will be clobbered when the file is regenerated.
+//
+//     Please read more about how to change this file in
+//     .github/CONTRIBUTING.md.
+//
+// ----------------------------------------------------------------------------
+
```
Comment (melinath) [inline mmv1/third_party/cai2hcl/services/compute/compute_forwarding_rule_test.go:14]:
```suggestion
```

File: mmv1/third_party/cai2hcl/converter_test.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
+package cai2hcl
+
+import (
+	"testing"
+
+	"github.com/GoogleCloudPlatform/terraform-google-conversion/v5/cai2hcl/services"
+	cai2hclTesting "github.com/GoogleCloudPlatform/terraform-google-conversion/v5/cai2hcl/testing"
+)
+
+func TestConvertCompute(t *testing.T) {
```
Comment (melinath) [inline mmv1/third_party/cai2hcl/converter_test.go:10]: Ideally these tests would live in the respective service packages - is that possible?

File: mmv1/third_party/cai2hcl/common/utils.go [hunk 4]
```diff
@@ -55,111 +56,61 @@ func MapToCtyValWithSchema(m map[string]interface{}, s map[string]*schema.Schema
...
-	tpgProvider := tpg.Provider()
-
-	result := make(map[string]Converter, len(converterFactories))
-	for name, factory := range converterFactories {
-		result[name] = factory(name, tpgProvider.ResourcesMap[name].Schema)
-	}
-
-	return result
+func NewConfig() *tpg_transport.Config {
+	// Currently its not needed, but it may change in future.
```
Comment (melinath) [inline mmv1/third_party/cai2hcl/common/utils.go:60]: Could we add this once it's needed instead of now?

File: mmv1/third_party/cai2hcl/common/uber_converter.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
+package common
+
+import (
+	"github.com/GoogleCloudPlatform/terraform-google-conversion/v5/caiasset"
+)
+
+// Converter which aggregates all service-specific converters in the same interface.
+type UberConverter struct {
```
Comment (melinath) [inline mmv1/third_party/cai2hcl/common/uber_converter.go:8]: I would be interested in potentially calling this `Converter` and having a `NewConverter` function, so that the implementation more closely matches tfplan2cai: https://github.com/GoogleCloudPlatform/terraform-google-conversion/blob/main/tfplan2cai/converters/google/convert.go

That said, I'm also open to alternatives - mostly I think it would be good to converge on a more similar implementation between the two halves over time so that we can reduce cognitive load moving between them.

What would you think?

File: mmv1/third_party/cai2hcl/common/hclwrite.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/hcl/hcl/printer"
+	"github.com/hashicorp/hcl/v2/hclwrite"
+	"github.com/zclconf/go-cty/cty"
+)
+
+func HclWriteBlocks(blocks []*HCLResourceBlock) ([]byte, error) {
```
Comment (melinath) [inline mmv1/third_party/cai2hcl/common/hclwrite.go:11]: The HCL & converter_factory extraction is extremely straightforward but ends up causing a decent chunk of the diff and making it harder to track the remaining changes - could you move that to a separate PR that I can approve and merge while we discuss the converter interface?

File: mmv1/third_party/cai2hcl/common/uber_converter.go (hunk 5 above)
Comment (melinath) [inline mmv1/third_party/cai2hcl/common/uber_converter.go:8]: What is the purpose of this refactor? Why is it necessary for the upcoming changes?

**Review by melinath — CHANGES_REQUESTED** [review]

File: mmv1/third_party/cai2hcl/converter_test.go (hunk 3 above)
Comment (amirkaromashkin) [reply mmv1/third_party/cai2hcl/converter_test.go:10]: This already works like that - each converter has its own test and uses local UberConverter to execute the conversion.

This test is a glue test to check the integration point between services

File: mmv1/third_party/cai2hcl/common/hclwrite.go (hunk 6 above)
Comment (amirkaromashkin) [reply mmv1/third_party/cai2hcl/common/hclwrite.go:11]: Sure, done: https://github.com/GoogleCloudPlatform/magic-modules/pull/9607

File: mmv1/third_party/cai2hcl/common/uber_converter.go (hunk 5 above)
Comment (amirkaromashkin) [reply mmv1/third_party/cai2hcl/common/uber_converter.go:8]: In Cai2hcl Converter is an [interface](https://github.com/GoogleCloudPlatform/terraform-google-conversion/blob/main/cai2hcl/common/converter.go), so we cannot call it so.

UberConverter is on one hand implements this Converter interface, but on the other - itself a new interface for group of converters... The naming is frustrating, I agree.

Comment (amirkaromashkin) [reply mmv1/third_party/cai2hcl/common/uber_converter.go:8]: The main reason was to incorporate new converter matching strategy - "by asset selfLink RegExp". This is needed to map ComputeForwardingRule asset type to either "compute_forwarding_rule" or "global_compute_forwarding_rule" converter based on a selfLink regexp. In Compute API both of resources have the same type - this is a common practice in Compute.

This new strategy will add 3rd property in the converter_map.go "interface", while there is no such interface in code. So the main purpose was to create a single interface for service-level converters aggregator.

Comment (amirkaromashkin) [reply mmv1/third_party/cai2hcl/common/uber_converter.go:8]: Here is an example of a matcher implementation: https://github.com/amirkaromashkin/magic-modules/blob/feature/cai2hcl_provider_generators/mmv1/third_party/cai2hcl/services/compute/converter.go

**Comment by melinath** [comment]
Going through and reminding myself of some background and noting it here for myself / posterity. This PR is a follow-up to https://github.com/GoogleCloudPlatform/terraform-google-conversion/pull/1416

File: mmv1/third_party/cai2hcl/common/hcl_write.go [hunk 7]
```diff
@@ -3,10 +3,25 @@ package common
 import (
 	"fmt"
 
+	"github.com/hashicorp/hcl/hcl/printer"
 	"github.com/hashicorp/hcl/v2/hclwrite"
 	"github.com/zclconf/go-cty/cty"
 )
 
+func HclWriteBlocks(blocks []*HCLResourceBlock) ([]byte, error) {
```
Comment (melinath) [inline mmv1/third_party/cai2hcl/common/hcl_write.go:11]: Splitting this out seems like a good idea! 👍 I sort of wish we had a name other than HCLResourceBlock for the input since it's not actually an HCL-specific format, just an intermediate container... but that's not something to change right now.

I'd be fine to approve and merge this change in a separate PR if you want but it's not really impacting readability.

File: mmv1/third_party/cai2hcl/common/uber_converter.go (hunk 5 above)
Comment (melinath) [reply mmv1/third_party/cai2hcl/common/uber_converter.go:8]: Yeah, the interface would need to be renamed as well - or we would need to stop using interfaces like this. Overall I'm concerned that there's a lot of complexity / nesting going on that ends up obfuscating what's happening, and I'm not sure it's benefitting us. But I could totally be forgetting or not considering something!

FWIW my inclination would be that we standardize towards how tfplan2cai works (having a [ResourceConverter struct](https://github.com/GoogleCloudPlatform/terraform-google-conversion/blob/ad14a1096e314d4245d5467fe04ab302bea97630/tfplan2cai/converters/google/resources/services/compute/compute_forwarding_rule.go#L50) with a predictable structure - this is similar to how [Terraform resources are implemented](https://github.com/hashicorp/terraform-provider-google/blob/20615b0badd5121b90878658b185c62a6dd64416/google/services/compute/resource_compute_forwarding_rule.go#L53).

This would basically mean replacing the New*Converter functions with a `ResourceConverter` that returns a standard type which knows the Terraform type it converts to.

We could than have a single map that of all asset types to relevant resource converters, similar to [resource_converters.go](https://github.com/GoogleCloudPlatform/terraform-google-conversion/blob/0ff3c33174570270e0bc937ebadae0b6ee80d6ed/tfplan2cai/converters/google/resources/resource_converters.go#L72) in tfplan2cai and [provider_mmv1_resources.go](https://github.com/hashicorp/terraform-provider-google/blob/20615b0badd5121b90878658b185c62a6dd64416/google/provider/provider_mmv1_resources.go#L123) in the provider. It's not elegant but it's very clear what's going on, and we would be able to avoid the indirection of the ConverterMap (and I think the names of things could be clearer as well.)

I have a vague memory that there was a reason we didn't do this previously - possibly a circular dependency? - but I don't remember why it was a problem for cai2hcl and not for tfplan2cai. And that might have been related to a completely different decision.

As before, I'm open to alternatives (and even hypothetically keeping the current structure) if there's a clear benefit or a blocker of some kind that outweighs standardizing on our other libraries.

File: mmv1/third_party/cai2hcl/common/utils.go [hunk 8]
```diff
@@ -39,10 +36,14 @@ func DecodeJSON(data map[string]interface{}, v interface{}) error {
 
 // Converts resource from untyped map format to TF JSON.
 func MapToCtyValWithSchema(m map[string]interface{}, s map[string]*schema.Schema) (cty.Value, error) {
+	// Normalize non-marshable properties manually.
+	m = normalizeFlattenedMap(m, s).(map[string]interface{})
```
Comment (melinath) [inline mmv1/third_party/cai2hcl/common/utils.go:40]: What is this replacing? I have a vague memory of needing to do this but I don't see anything eqivalent that was removed at a quick scan.

**Review by melinath — CHANGES_REQUESTED** [review]
I know that you've been waiting for review on this for a while and I'm sorry about the delay and that I'm asking about refactoring further - but since you're proposing changes and we're moving from "proof of concept" towards what we expect to have in place longer-term it would be great to be moving towards what might be a more final design.

**Comment by melinath** [comment]
@iyabchen since you did the initial implementation it would be great if you could take a look and add any relevant context that you might remember.

File: mmv1/third_party/cai2hcl/common/utils.go (hunk 8 above)
Comment (amirkaromashkin) [reply mmv1/third_party/cai2hcl/common/utils.go]: Oh, right, its not needed yet, because current there are no yet generated converters, this will be needed later, removed

**Comment by amirkaromashkin** [comment]
Following this thread: https://github.com/GoogleCloudPlatform/magic-modules/pull/9378#discussion_r1421035897 (not sure why I cant just respond to it directly).

There are 2 topics, lets discuss them separately.

1. ConverterMap location (single file of 1 file per API folder): 
Your proposal is to have a single ConverterMap file for all converters, not a file for per feature area.
One of the issues I see is indeed a cycle dep sue to unit tests, which are in cai2chl located near the converters in cai2hcl, in tfplan2cai - outside. This would create a cycle cai2hcl -(converter ref)-> compute -(converterMap ref from unit test)-> cai2hcl. Still I can try to overcome somehow (by extracting some new package maybe).

2. ConverterMap structure.
tfplan2cai is a mapping from string to ResourceConverter. But in Cai2hcl, as I mentioned in comments above ([here](https://github.com/GoogleCloudPlatform/magic-modules/pull/9378#discussion_r1418164466)) there is a different chain: Asset Type -> matchingStrategy -> ResourceConverter.
Its because I.e. compute_instance_template and compute_region_instance_template they have the same Asset Type: InstanceTemplate ([CAI Docs](https://cloud.google.com/asset-inventory/docs/supported-asset-types)).
Same applies to  InstanceGroup, ForwardingRule, BackendService and many other Compute types.
For FR and BS its not 100% so because CAI supported types list this "artificial" types like GlobalForwardingRule and RegionBackendService, still it needs additional effort to set these types in our codebase., Also - its officially not the case InstanceGroup and InstanceTemplate, so I would prefer to not mandate this differentiator from library clients and use a self link matcher.
We can still refer to this file as converter_map.go though, but the value should not be the converter, but smth else.
What I mean through all of this, is that ConverterMap is more complicated in cai2hcl than in tfplan2cai and we should not require too much similarity to their interface/structure.
Also that being said I'm not so sure that 1 file will work considering the increased complexity of the mapping.

So to sum up - I will try to:
1. Have 1 converter_map file totally
2. Stick to the converter_map pattern, while keeping the converter matching flexibility (by selflink pattern)

**Comment by amirkaromashkin** [comment]
> Following this thread: [#9378 (comment)](https://github.com/GoogleCloudPlatform/magic-modules/pull/9378#discussion_r1421035897) (not sure why I cant just respond to it directly).
> 
> There are 2 topics, lets discuss them separately.
> 
> 1. ConverterMap location (single file of 1 file per API folder):
>    Your proposal is to have a single ConverterMap file for all converters, not a file for per feature area.
>    One of the issues I see is indeed a cycle dep sue to unit tests, which are in cai2chl located near the converters in cai2hcl, in tfplan2cai - outside. This would create a cycle cai2hcl -(converter ref)-> compute -(converterMap ref from unit test)-> cai2hcl. Still I can try to overcome somehow (by extracting some new package maybe).
> 2. ConverterMap structure.
>    tfplan2cai is a mapping from string to ResourceConverter. But in Cai2hcl, as I mentioned in comments above ([here](https://github.com/GoogleCloudPlatform/magic-modules/pull/9378#discussion_r1418164466)) there is a different chain: Asset Type -> matchingStrategy -> ResourceConverter.
>    Its because I.e. compute_instance_template and compute_region_instance_template they have the same Asset Type: InstanceTemplate ([CAI Docs](https://cloud.google.com/asset-inventory/docs/supported-asset-types)).
>    Same applies to  InstanceGroup, ForwardingRule, BackendService and many other Compute types.
>    For FR and BS its not 100% so because CAI supported types list this "artificial" types like GlobalForwardingRule and RegionBackendService, still it needs additional effort to set these types in our codebase., Also - its officially not the case InstanceGroup and InstanceTemplate, so I would prefer to not mandate this differentiator from library clients and use a self link matcher.
>    We can still refer to this file as converter_map.go though, but the value should not be the converter, but smth else.
>    What I mean through all of this, is that ConverterMap is more complicated in cai2hcl than in tfplan2cai and we should not require too much similarity to their interface/structure.
>    Also that being said I'm not so sure that 1 file will work considering the increased complexity of the mapping.
> 
> So to sum up - I will try to:
> 
> 1. Have 1 converter_map file totally
> 2. Stick to the converter_map pattern, while keeping the converter matching flexibility (by selflink pattern)

Done.
1. Renamed UberConverter to ConverterMap.
2. Removed ConverterMap's from each API folder and replace them with a single top-level ConverterMap
3. Changed unit tests packages to "<API_NAME>_test" to avoid dependency cycle.

Some visible potential issues are:
1. Unit tests packages are different then the folder name.
2. ConverterMap is not actually a "map", but a configuration object with maps (now 2 maps, in future - another Matchers map will come).
IMO, these compromises are ok.

**Review by melinath — CHANGES_REQUESTED** [review]
(The reason you couldn't reply to the comment was because it was technically a comment on an existing thread, and GitHub threading is weird and confusing and will only let you comment on the main copy of a particular thread.)

Using *_test packages should be fine. 👍 

I put down some more thoughts about the structure here - let me know what you think!

File: mmv1/third_party/cai2hcl/common/converter_map.go [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+type ConverterMap struct {
+	// Mapping from CAI Asset Type to converter name.
+	AssetTypeToConverter map[string]string
+
+	// Mapping from converter name and converter.
+	Converters map[string]Converter
+}
+
+// ConvertMap implements Converter interface to be able to convert any Asset type.
+func (c ConverterMap) Convert(assets []*caiasset.Asset) ([]*HCLResourceBlock, error) {
```
[the remark this answers is not in the document]
Comment (iyabchen) [reply mmv1/third_party/cai2hcl/common/converter_map.go:17]: If I understand this correctly, it means
grouping used to be 
- 'google_project' group: project1, project2, billing1
- 'google_network' group: network1, network2

And your comment mentioned this should be changed to grouping like this?  probably need to mix the name + type together as the hash key. 
- project1, billing1
- project2
- network1
- network2

I think it makes sense, and the difference is the same type of converter will be initiated multiple times rather than 1 time in the current style of grouping. 

---
`tpg_provioder.Provider()` can be stored as a global variable if the schema is required at multiple invocation of converter init.

---

Merging like the tf->cai way is not that easy in this context. mainly because cty.Value is not that easy to manipulate like json.

[the remark this answers is not in the document]
Comment (amirkaromashkin) [reply mmv1/third_party/cai2hcl/common/converter_map.go:17]: It does make sense. It all sounds legit.

One thing I don't yet know is how to implement getName(asset Asset) { ... } function in each converter, when we think of it from generator perspective.

I.e. I planned to expose AssetType and SelfLinkRegExp like [this](https://github.com/amirkaromashkin/magic-modules/blob/main-cai2hcl/mmv1/templates/cai2hcl/resource_converter.go.erb#L43). 

I will continue looking how to do that.

[the remark this answers is not in the document]
Comment (amirkaromashkin) [reply mmv1/third_party/cai2hcl/common/converter_map.go]: As discussed offline with Stephen - there is no need to group assets by asset name due to current converters architecture.

PTAL - I have just significantly simplified the overall architecture: 
1. Got rid of multi-layered converter maps => all converters are installed in a single file converter_map
2. Got rid of UberConverters => one converter_map is left.
2. Got rid of converter factories
3. Moved converter names constants inside the Converters (not sure if this is a good choice)

I tried to simplify it even further and converter converter_map to converters array ([here](https://github.com/amirkaromashkin/terraform-google-conversion/commit/c8833078c01087e333bd36cef5b23ac4a03bb21b#diff-f25bf0a8d306a73f506a5e38af63497535b086d6e1a79d201fb1091e5a2007bb)), but I believe we current approach is actually a better one (clearly states which converters affect which asset types)

One thing I'm not sure is wether it is really needed to move Converter Name inside the converters: current converter names are exactly TF schemas, but its not necessarily so. So leaving them as raw strings, I think was a fine choice either.

File: mmv1/third_party/cai2hcl/common/converter_map.go [hunk 10]
```diff
@@ -0,0 +1,44 @@
+package common
+
+import (
+	"github.com/GoogleCloudPlatform/terraform-google-conversion/v5/caiasset"
+)
+
+// Configuration object to map CAI Assets to Converters.
```
[the remark this answers is not in the document]
Comment (amirkaromashkin) [reply mmv1/third_party/cai2hcl/common/converter_map.go]: Oh, right, thank you, I did not follow that practice - thanks, fixed

**Review by melinath — CHANGES_REQUESTED** [review]
Overall this is looking good. I'm still concerned about the maintenance burden / cognitive load of multiple maps for converters; I have a suggestion that I think should address it - let me know what you think!

File: mmv1/third_party/cai2hcl/convert.go [hunk 11]
```diff
@@ -21,14 +21,32 @@ func Convert(assets []*caiasset.Asset, options *Options) ([]byte, error) {
...
+	// Group resources from the same tf resource type for convert.
+	// tf -> cai has 1:N mappings occasionally
+	groups := make(map[string][]*caiasset.Asset)
+	for _, asset := range assets {
+
+		name, _ := ConverterMap.AssetTypeToConverterName[asset.Type]
+		if name != "" {
+			groups[name] = append(groups[name], asset)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (amirkaromashkin) [reply mmv1/third_party/cai2hcl/convert.go]: I'd prefer not to use converter object as a hash key: i.e. currently converter instance has fields of type Map (which make it non-hashable) and I would not want to mandate owners to keep converters hashable.
So I think we do need some converter ID.

Still I have slightly simplified the converter_map, PTAL.

**Review by melinath — APPROVED** [review]
I'm not entirely satisfied with the converter map structure but I may be overthinking it. It's much easier to understand as-is and should be fairly easy to refactor in the future if needed (and we might know more than we do now.) This LGTM.
